src/Extensions.Hosting.AsyncInitialization/Hosting/AsyncInitializationHostExtensions.cs

```diff
@@ -18,19 +18,68 @@ public static class AsyncInitializationHostExtensions
         /// <param name="host">The host.</param>
         /// <param name="cancellationToken">Optionally propagates notifications that the operation should be cancelled</param>
         /// <returns>A task that represents the initialization completion.</returns>
+        /// <exception cref="ArgumentNullException">Thrown when the host is null.</exception>
+        /// <exception cref="InvalidOperationException">Thrown when the initialization service has not been registered.</exception>
+        /// <exception cref="OperationCanceledException">Thrown when the cancellationToken is cancelled.</exception>
         public static async Task InitAsync(this IHost host, CancellationToken cancellationToken = default)
         {
-            if (host == null)
-                throw new ArgumentNullException(nameof(host));
+            if (host == null) throw new ArgumentNullException(nameof(host));
 
-            using var scope = host.Services.CreateScope();
-            var rootInitializer = scope.ServiceProvider.GetService<RootInitializer?>();
-            if (rootInitializer == null)
+            await using var scope = host.Services.CreateAsyncScope();
+            var rootInitializer = scope.ServiceProvider.GetService<RootInitializer>()
+                ?? throw new InvalidOperationException("The async initialization service isn't registered, register it by calling AddAsyncInitialization() on the service collection or by adding an async initializer.");
+
+            await rootInitializer.InitializeAsync(cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Performs teardown, by calling all registered async initializers that implement <see cref="IAsyncTeardown"/>
+        /// </summary>
+        /// <param name="host">The host.</param>
+        /// <param name="cancellationToken">Optionally propagates notifications that the operation should be cancelled</param>
+        /// <returns>The <see cref="Task"/> that represents the asynchronous operation.</returns>
+        /// <exception cref="ArgumentNullException">Thrown when the host is null.</exception>
+        /// <exception cref="InvalidOperationException">Thrown when the initialization service has not been registered.</exception>
+        /// <exception cref="OperationCanceledException">Thrown when the cancellationToken is cancelled.</exception>
+        private static async Task TeardownAsync(this IHost host, CancellationToken cancellationToken = default)
+        {
+            if (host == null) throw new ArgumentNullException(nameof(host));
+
+            await using var scope = host.Services.CreateAsyncScope();
+            var rootInitializer = scope.ServiceProvider.GetService<RootInitializer>()
+                ?? throw new InvalidOperationException("The async initialization service isn't registered, register it by calling AddAsyncInitialization() on the service collection or by adding an async initializer.");
```

Comment:
This is already checked in `InitAndRunAsync`, and this method is only called from there, so we don't need that check.

Also, you're creating a new scope and resolving `RootInitializer` again in that new scope, which means it will be a new instance, with new instances of the initializers (unless they're singletons).
I think you should use the same instance of `RootInitializer` as for initialization. At this point, you might as well remove the whole `TeardownAsync` method, and just call `rootInitializer.TeardownAsync` directly in `InitAndRunAsync`